### PR TITLE
field preparation functions

### PR DESCRIPTION
**Pull request by soniktrooth**
Image fields, 
text fields,
Entity ref,
link handling

### CODE REVIEW

**Comment by soniktrooth** [comment]
yo @derekderaps I'd love to talk about this stuff and see if you think this is going in a good direction

File: kalastatic.module [hunk 1]
```diff
@@ -73,3 +73,93 @@ function kalastatic_preprocess_html(&$variables) {
     $variables['build_path'] = base_path() . 'kalastatic/';
   }
 }
+
+/**
+* Prepare the variables from a text_with_summary field.
+ */
+function kalastatic_prepare_field__text_with_summary($field, $entity) {
+  $output = [];
+  foreach ($entity->$field->getValue() as $i => $item) {
```
Comment (hawkeyetwolf) [inline kalastatic.module:82]: Text w/summary might be one of the easier places to try using the rendered output. We may even want to consider adding a nested layer to the template data to support `text.value` and `text.summary` or something like that.

File: kalastatic.module [hunk 2]
```diff
@@ -73,3 +73,93 @@ function kalastatic_preprocess_html(&$variables) {
...
+
+/**
+ * Prepare the variables from an image field, ready for the image.html.twig
+ */
+function kalastatic_prepare_field__entity_reference($field, $entity, $formatter) {
+  $output = [];
+  foreach ($entity->$field->referencedEntities() as $i => $item) {
+    if ($formatter) {
+      // Caller wants the referenced entites in a specific format.
+      // TODO: Can we use actual field formatters/view modes somehow?
```
Comment (hawkeyetwolf) [inline kalastatic.module:120]: Totes I think we could. We could look up the field formatter (+ its configuration) chosen for the given bundle and then use it to render the entity. This is cool but I'm fine with postponing it for now. Note to self for later, we'd have to assume a view mode to use; we could either just go with "full" (or whatever the D8 version of "full" is) or always use a custom view mode like "kalastatic" where we configure the field formatters for how we want them to render in the template.

**Review by hawkeyetwolf — APPROVED** [review]
Cool beans, duderino! I have not given it the most careful eye, but I definitely approve as "works for now" approach and we can work on leveraging field formatters and standard drupal render pipelines at a later date. Sharp work, aussieman!

**Comment by hawkeyetwolf** [comment]
The only other question I had was where these functions were getting called from. Perhaps that's coming later.

File: kalastatic.module (hunk 1 above)
Comment (soniktrooth) [reply kalastatic.module:82]: text.value/text.summary is a good idea but it breaks our pre-established convention with text though. I guess we may end up with some extra layers in our json data on the prototype side but it's probably worth it to ensure compatibility with the Derpals

File: kalastatic.module (hunk 2 above)
Comment (soniktrooth) [reply kalastatic.module:120]: I really like the idea of having a 'kalastatic' view mode. That would make it way more obvious in the UI that a particular entity is using KS and not be as mysterious for a non-kalabot.

**Comment by soniktrooth** [comment]
@derekderaps These functions are being called from `kalablog_preprocess_node(&$vars)` in this PR: https://github.com/kalamuna/kalablog8/pull/129 but we're going to want to merge this first otherwise that will fail.

**Comment by soniktrooth** [comment]
Also, regarding "leveraging field formatters and standard drupal render pipelines" I totally agree. It was more straight forward in D7 because we could just yank the `$vars['content']` and traverse the array but D8 doesn't seem to have every thing there for us to just cherry pick so we need to learn more about the render pipeline and try to make sure we're not bypassing any Drupal stuffs in our approach.

**Pull request merged**